components/list/list-item-drag-handle.js

```diff
@@ -0,0 +1,204 @@
+				width: 0.9rem;
+			}
+			/* Firefox includes a hidden border which messes up button dimensions */
+			button::-moz-focus-inner {
+				border: 0;
+			}
+			.d2l-list-item-drag-handle-dragger-button {
+				background-color: unset;
+			}
+			button,
+			button[disabled]:hover,
+			button[disabled]:focus {
+				background-color: var(--d2l-color-gypsum);
+				color: var(--d2l-color-ferrite);
+			}
+			button:hover,
+			button:focus {
+				background-color: var(--d2l-color-mica);
+			}
+			button[disabled] {
+				cursor: default;
+				opacity: 0.5;
+			}
+		`];
+	}
+
+	constructor() {
+		super();
+		this._keyboardActive = false;
+		this.disabled = false;
+	}
+
+	focus() {
+		const node = getFirstFocusableDescendant(this);
+		if (node) node.focus();
+	}
+
+	render() {
+		return html`${this._keyboardActive && !this.disabled ? this._renderKeyboardDragging() : this._renderDragger()}`;
+	}
+
+	updated(changedProperties) {
+		if (changedProperties.has('_keyboardActive')) {
+			this.focus();
+		}
+	}
+
+	_dispatchAction(action) {
+		if (!action) return;
+		this.dispatchEvent(new CustomEvent('d2l-list-item-drag-handle-action', {
+			detail: { action },
+			bubbles: false
+		}));
+	}
+
+	_dispatchActionDown() {
+		this._dispatchAction(dragActions.down);
+	}
+
+	_dispatchActionUp() {
+		this._dispatchAction(dragActions.up);
+	}
+
+	_onBlur() {
+		this._keyboardActive = false;
+		if (!this._tabbing) {
+			this._dispatchAction(dragActions.save);
+		}
+	}
+
+	_handleActiveKeyboard(e) {
+		if (!this._keyboardActive) {
+			return;
+		}
+		let action = null;
+		switch (e.key) {
+			case UP:
+				action = dragActions.up;
+				break;
+			case DOWN:
+				action = dragActions.down;
+				break;
+			case HOME:
+				action = dragActions.first;
+				break;
+			case END:
+				action = dragActions.last;
+				break;
+			case TAB:
+				action = e.shiftKey ? dragActions.previousElement : dragActions.nextElement;
+				break;
+			case ESC:
+				action = dragActions.cancel;
+				this.updateComplete.then(() => this._keyboardActive = false);
+				break;
+			case ENTER:
+			case SPACE:
+			case RIGHT:
+				action = dragActions.save;
+				this.updateComplete.then(() => this._keyboardActive = false);
+				break;
+			default:
+				return;
+		}
+		this._dispatchAction(action);
+		e.preventDefault();
+	}
+
+	_handleInactiveKeyboard(e) {
+		if (e.type === 'click' || e.key === ENTER || e.key === SPACE || e.key === LEFT) {
+			this._dispatchAction(dragActions.active);
+			this._keyboardActive = true;
+			e.preventDefault();
+		}
+	}
+
+	_handlePreventDefault(e) {
+		e.preventDefault();
+	}
+
+	_handleInactiveKeyDown(e) {
+		if (e.type === 'click' || e.key === ENTER || e.key === SPACE || e.key === LEFT) {
+			e.preventDefault();
+		}
+	}
+
+	_renderDragger() {
+		return html`
+			<button
+				class="d2l-list-item-drag-handle-dragger-button"
+				@click="${this._handleInactiveKeyboard}"
+				@keyup="${this._handleInactiveKeyboard}"
+				@keydown="${this._handleInactiveKeyDown}"
```

Comment:
why keydown handler is different? Does it have special meaning when user presses key DOWN?

Reply:
Yes, so `keydown` happens when a key is being pressed. The reason for this being different is because I want all keys but the ones I am defining on `keyup` to work as normal. However, I need to prevent the keys in the `keyup` to not doing anything since `enter` also activates from the `click` event. So I get a double activation which makes it flash back and forth.